File: tools/linux/getkcore.py
"""
Author: Andrew Case / andrew@dfir.org
License: GPLv2

TOOLS PURPOSE:
64-bit Linux Physical Memory Acquistion from Userland

NOT FOR PUBLIC RELEASE:
This file is not to be distributed publicly until the release of the Art of Memory Forensics is published
A cleaned up version of it will be released with the book's materials

ACQUISTION ALGORITHM:
This script relies on the static virtual mapping of all RAM kept by x64 Linux systems.
This mapping is illustrated here: http://lxr.free-electrons.com/source/Documentation/x86/x86_64/mm.txt

To reach these mappings we use the /proc/kcore file.
This file exposes all of physical memory (including hardware devices) as ELF sections of a core dump file.

To acquire memory, the script first parses /proc/iomem and determines ranges of "System RAM".
It then parses the sections of /proc/kcore and matches "System RAM" regions to those found in the kcore file.
This matching is possible by using the static offset (0xffff880000000000) of the virtual mapping of RAM.
See the _find_kcore_sections function for this algorithm

Each RAM region found is then written to a LiME formatted file so that it can be immediately analyzed with Volatility.

"""

import os, sys, struct

try:
    import elftools
except:
    print "pyelftools not found. Please install"
    exit(1)

from elftools.elf.elffile import ELFFile

##### CHANGE THIS TO MAKE DEBUG MESSAGES PRINT DURING ACQUISTION #####
debug = 1

def _print_msg(msg):
    print msg

def _debug_msg(msg):
    if debug:
        _print_msg(msg)

def _die(msg):
    _print_msg(msg + " Exiting.")
    exit(1)

def _do_startup_checks():
    if sys.platform.find("linux") == -1:
        die("This script only acquires memory from Linux systems.")

    if os.getuid() != 0:
        die("This script must be run as root.")

    if not os.path.exists("/proc/kcore"):
        die("/proc/kcore not enabled on this system.") 

def _parse_proc_iomem():
    ranges = {}

    ram_regions = [x for x in open("/proc/iomem", "r").readlines() if x.find("System RAM") != -1]

    # 00010000-0009efff : System RAM\n
    for region in ram_regions:
        # gets the addresses
        first = region.split()[0]

        ents = first.split("-")

        (start, end) = (int(ents[0], 16), int(ents[1], 16))

        _debug_msg("_parse_proc_iomem: Adding range %#x-%#x" % (start, end))

        ranges[start + 0xffff880000000000] = (start, end)

    return ranges

def  _find_kcore_sections(ram_ranges):
    sections = []

    with open("/proc/kcore", 'rb') as f:
        elffile = ELFFile(f)

        for section in elffile.iter_segments(): 
            file_off = section['p_offset']
            vaddr    = section['p_vaddr']
            sz       = section['p_memsz']

            if not vaddr in ram_ranges:
                continue

            # the start address in physical memory of the system from /proc/iomem
            phys_off = ram_ranges[vaddr][0]

            _debug_msg("Found RAM region: %.08x %.08x %.08x" % (file_off, vaddr, sz))

            sections.append((file_off, phys_off, sz)) 
    
    return sections

def _make_lime_header(phys_off, sz):
    magic   = struct.pack("<I", 0x4C694D45)
    version = struct.pack("<I", 1)
    start   = struct.pack("<Q", phys_off)
    end     = struct.pack("<Q", phys_off + sz - 1)
    reserved = "\x00" * 8

    _debug_msg("Created lime header: %.16x %.16x" % (phys_off, phys_off + sz - 1))

    return magic + version + start + end + reserved     

def _write_file(outfd, data, size_of_data):    
    wrote = os.write(outfd, data)

    if wrote != size_of_data:
        _die("_write_file: Asked to write %d bytes but %d written!" % (size_of_data, wrote))

def _write_section_data(outfd, file_off, sz):
    chunk_size = 10000000

    kcore_fd = os.open("/proc/kcore", os.O_RDONLY)
    os.lseek(kcore_fd, file_off, os.SEEK_SET)

    wrote = 0

    while wrote < sz:
        left = sz - wrote

        if left < chunk_size:
            to_read = left
        else:   
            to_read = chunk_size

        buf = os.read(kcore_fd, to_read)
        
        buf_wrote = os.write(outfd, buf)

        if buf_wrote != to_read:
            _die("_write_section_data: Attempt to read from %d bytes from %x read %d bytes" % (to_read, file_off + wrote, to_read))

        wrote = wrote + to_read

    if wrote != sz:
        _die("_write_section_data: Should have acquired %d bytes from section %x but instead acquired %d" % (sz, file_off, wrote))

    os.close(kcore_fd)

def _write_memory_dump(outfd, kcore_sections):
    for (file_off, phys_off, sz) in kcore_sections:
        lime_header = _make_lime_header(phys_off, sz)

        _write_file(outfd, lime_header, len(lime_header))

        _write_section_data(outfd, file_off, sz)

# entry point to acquistion process
def create_memory_dump(output_file):
    _do_startup_checks()

    _debug_msg("Enumerating ranges from /proc/iomem")
    ram_ranges = _parse_proc_iomem()

    _debug_msg("Enumerating sections from /proc/kcore")
    kcore_sections = _find_kcore_sections(ram_ranges)
    
    outfd = os.open(output_file, os.O_WRONLY|os.O_CREAT, 770)

    _debug_msg("Writing sections to LiME formatted file")
    _write_memory_dump(outfd, kcore_sections)

    os.close(outfd)

def main(): 
    if len(sys.argv) < 2:
        die("Usage: %s <output file name>" % (sys.argv[0]))
    
    outfile = sys.argv[1]

    create_memory_dump(outfile)


if __name__ == "__main__":
    main()

File: tools/linux/kcore/getkcore.c
/*

Author: Andrew Case / andrew@dfir.org
License: GPLv2

TOOLS PURPOSE:
64-bit Linux Physical Memory Acquistion from Userland

NOT FOR PUBLIC RELEASE:
This file is not to be distributed publicly until the release of the Art of Memory Forensics is published
A cleaned up version of it will be released with the book's materials

ACQUISTION ALGORITHM:
This script relies on the static virtual mapping of all RAM kept by x64 Linux systems.
This mapping is illustrated here: http://lxr.free-electrons.com/source/Documentation/x86/x86_64/mm.txt

To reach these mappings we use the /proc/kcore file.
This file exposes all of physical memory (including hardware devices) as ELF sections of a core dump file.

To acquire memory, the script first parses /proc/iomem and determines ranges of "System RAM".
It then parses the sections of /proc/kcore and matches "System RAM" regions to those found in the kcore file.
This matching is possible by using the static offset (0xffff880000000000) of the virtual mapping of RAM.
See the _find_kcore_sections function for this algorithm

Each RAM region found is then written to a LiME formatted file so that it can be immediately analyzed with Volatility.

*/

#define _LARGEFILE64_SOURCE

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <fcntl.h>
#include <stdarg.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/types.h>

#include "elf.h"
#include "getkcore.h"

static int debug = 0;

// how much data is read at once from /proc/kcore
static unsigned int chunk_size = 10000000;

void _debug_msg(const char *format,...)
{
    if (debug)
    {
        va_list va;
        va_start(va,format);
        vfprintf(stderr,format,va);
        va_end(va);
        printf("\n");
    }
}

void _die(const char* format,...) 
{
    va_list va;
    va_start(va,format);
    vfprintf(stderr,format,va);
    va_end(va);
    printf("\n");
    
    exit(1);
}

void _do_startup_checks(void)
{
    if (getuid() != 0)
        _die("This program must be run as root");

    if (access("/proc/kcore", F_OK) == -1)
        _die("/proc/kcore does not exist");  
}

void _write_lime_header(int out_fd, unsigned long long phys_off, unsigned long long size)
{
    lime_range l;

    l.magic   = 0x4C694D45;
    l.version = 1; 
    l.s_addr  = phys_off;
    l.e_addr  = phys_off + size - 1;
    memset(&l.reserved, 0x00, sizeof(l.reserved));

    _debug_msg("_write_lime_header: Made lime header for start: %llx end: %llx", l.s_addr, l.e_addr);

    if (write(out_fd, &l, sizeof(l)) != sizeof(l))
        _die("_write_lime_header: Error writing header for offset: %x", phys_off);
} 

void _read_write_region(int kcore_fd, int out_fd, Elf64_Phdr *p, unsigned long long phys_start, unsigned char *read_buf)
{
    unsigned long long wrote;
    unsigned long long left;
    unsigned long long to_read;
    unsigned long long rw_sz;

    // seek to the offset where the region is
    if (lseek64(kcore_fd, p->p_offset, 0) != (off_t)p->p_offset)
        _die("_read_write_region: Unable to seek to file offset %llx", p->p_offset); 

    wrote = 0;

    // read & write the region
    while (wrote < p->p_memsz)
    {
        memset(read_buf, 0x00, chunk_size);
        
        left = p->p_memsz - wrote;

        if (left < chunk_size)
            to_read = left;
        else
            to_read = chunk_size;

        rw_sz = read(kcore_fd, read_buf, to_read);

        if (rw_sz != to_read)
            _die("_read_write_region: Requested to read %llx bytes from %llx | %llx but received %llx", to_read, phys_start, phys_start + wrote, rw_sz); 

        rw_sz = write(out_fd, read_buf, to_read);

        if (rw_sz != to_read)
            _die("_read_write_region: Requested to write %llx bytes from %llx | %llx but wrote %llx", to_read, phys_start, phys_start + wrote, rw_sz); 

        wrote = wrote + to_read;
    }
    
    printf("Wrote %llu bytes from %llx\n", wrote, phys_start);

}

void _process_header(int kcore_fd, int out_fd, unsigned long long phdr_addr, unsigned long long phys_start, unsigned char *read_buf)
{
    Elf64_Phdr p;
 
    if (lseek64(kcore_fd, phdr_addr, 0) != (off_t)phdr_addr)
        _die("_process_header: Unable to seek to program header's offset: %x", phdr_addr);

    if (read(kcore_fd, &p, sizeof(p)) != sizeof(p))
        _die("_process_header: Unable to read program header: %x | %x\n", phdr_addr, phys_start);

    if (phys_start + 0xffff880000000000 == p.p_vaddr)
    {
        _write_lime_header(out_fd, phys_start, p.p_memsz);
        _read_write_region(kcore_fd, out_fd, &p, phys_start, read_buf);
    }       
}
 
void _write_region(int kcore_fd, int out_fd, unsigned long long phys_start, unsigned char *read_buf)
{
    Elf64_Ehdr h;
    unsigned short i;

    if (lseek64(kcore_fd, 0, 0) != 0)
        _die("_write_region: Unable to seek to offset 0");

    if (read(kcore_fd, &h, sizeof(h)) != sizeof(h))
        _die("_write_region: Unable to read ELF header for offset: %llx\n", phys_start);
 
    for (i = 0; i < h.e_phnum; i++)
        _process_header(kcore_fd, out_fd, h.e_phoff + (i * sizeof(Elf64_Phdr)), phys_start, read_buf);

}

char *_read_proc_iomem(void)
{
    int fd;
    off_t size;
    char *contents;

    fd = open("/proc/iomem", O_RDONLY);

    if (fd == -1)
        _die("_read_proc_iomem: Unable to open /proc/iomem");  

    size = 1000000;

    contents = malloc(size + 2);

    if (contents == NULL)
        _die("_read_proc_iomem: Unable to allocate buffer for reading /proc/iomem");

    *(contents + size + 1) = 0x00;

    if (read(fd, contents, size) < 1)
        _die("_read_proc_iomem: Unable to read /proc/iomem");

    close(fd);

    return contents;
}

// Parses /proc/iomem and calls _write_region with each found 
void _dump_ranges(int kcore_fd, int out_fd, unsigned char *read_buf)
{
    off_t size;
    off_t curoff;
    char *contents;
    char *cur;
    char *curn; 
    char *intbuf;
    char *dash;   
    unsigned long long start;
    unsigned long long end;

    contents = _read_proc_iomem();

    curoff = 0;

    size = strlen(contents);

    while (curoff < size)
    {
        // break up by newline
        cur = contents + curoff;
        curn = strstr(cur, "\n");

        if (curn == NULL)
            break;

        *curn = 0x00;

        // skip to next line if not RAM
        if (strstr(cur, "System RAM") == NULL)
        {
            curoff = curoff + curn - cur + 1; 
            continue;        
        }
        
        // 00100000-3fedffff : System RAM
        intbuf = strstr(cur, " ");
        dash = strstr(cur, "-");

        if (intbuf == NULL || dash == NULL || intbuf < dash)
            _die("parse_proc_iomem: Line broke parser: %s", cur);

        *dash   = 0x00;
        *intbuf = 0x00;

        start = strtoull(cur, NULL, 16);
        end   = strtoull(dash + 1, NULL, 16);

        _debug_msg("Found RAM at start: %llx end: %llx", start, end);

        _write_region(kcore_fd, out_fd, start, read_buf);
            
        curoff = curoff + curn - cur + 1; 
    }
}

int create_memory_dump(char *outfile)
{
    int kcore_fd;
    int out_fd;
    unsigned char *read_buf;

    read_buf = malloc(chunk_size);
    if (read_buf == NULL)
        _die("_create_memory_dump: Unable to allocate /proc/kcore read buffer");
 
    _do_startup_checks();

    kcore_fd = open("/proc/kcore", O_RDONLY);

    if (kcore_fd == -1)
        _die("create_memory_dump: Unable to open /proc/kcore for reading");

    out_fd = open(outfile, O_WRONLY|O_CREAT, 0700);

    _dump_ranges(kcore_fd, out_fd, read_buf);

    close(kcore_fd);
    close(out_fd);

    return 0;
}

int main(int argc, char **argv)
{
    if (argc < 2)
        _die("Usage: ./getkcore <output file name>");  

    create_memory_dump(argv[1]);
    
    return 0;
}

